File: mini-variational-principle/src/constrained.jl
# constrained.jl — 约束变分: 等周问题, Lagrange乘子法, 可变端点, Bolza问题
# 参考: Gelfand & Fomin Ch.3, Bliss "Calculus of Variations"

# ============================================================
# 等周约束
# ============================================================
"等周约束 Lagrange 乘子法: min J[y]=∫Fdx s.t. K[y]=∫Gdx=c"
function isoperimetric_lagrange_multiplier(F::Function, G::Function, c::Float64, a::Float64, b::Float64, ya::Float64, yb::Float64; lambda_guess=0.0)
    function constraint(lambda::Float64)
        Fe(x,y,yp)=F(x,y,yp)+lambda*G(x,y,yp)
        dx=(b-a)/500; y=zeros(501); yp=zeros(501)
        y[1]=ya; yp[1]=(yb-ya)/(b-a); K=0.0
        for i in 1:500; x=a+(i-1)*dx; K+=G(x,y[i],yp[i])*dx
            dFdy_val=(Fe(x+1e-6,y[i],yp[i])-Fe(x-1e-6,y[i],yp[i]))/(2e-6)
            yp[i+1]=yp[i]+dx*dFdy_val; y[i+1]=y[i]+dx*yp[i]
        end; K-c
    end
    lo,hi=-100.0,100.0; flo,fhi=constraint(lo),constraint(hi)
    for _ in 1:60
        abs(flo)<1e-8 && return lo; abs(fhi)<1e-8 && return hi
        mid=(lo+hi)/2; fmid=constraint(mid)
        flo*fmid<0 ? (hi=mid;fhi=fmid) : (lo=mid;flo=fmid)
    end; (lo+hi)/2
end

"Dido 问题 (最大面积给定周长): 圆是最优解, R=L/(2π)"
function dido_solution(L::Float64; n=500)
    R=L/(2pi); theta=range(0,2pi,length=n)
    R*cos.(theta), R*sin.(theta), R
end

# ============================================================
# 一般约束的 Lagrange 乘子法
# ============================================================
"一般 Lagrange 乘子泛函: L[x,y,y'']=F+Σλ_i G_i"
function lagrange_multiplier_functional(F::Function, constraints::Vector{Tuple{Function,Float64}}, a::Float64, b::Float64, ya::Float64, yb::Float64; lambda_init=nothing)
    m=length(constraints)
    lambdas=isnothing(lambda_init) ? zeros(m) : collect(lambda_init)
    # 简化: 返回 lambdas 满足所有约束
    lambdas
end

"单约束 BVP 求解"
function solve_constrained_el_bvp(F::Function, G::Function, c::Float64, a::Float64, b::Float64, ya::Float64, yb::Float64; n_grid=500, tol=1e-8)
    lambda=isoperimetric_lagrange_multiplier(F,G,c,a,b,ya,yb)
    Fe(x,y,yp)=F(x,y,yp)+lambda*G(x,y,yp)
    # 定义 dFdy,dFdyp,d2Fdyp2 为数值近似
    dFdy(x,y,yp)=(Fe(x+1e-6,y,yp)-Fe(x-1e-6,y,yp))/(2e-6)
    dFdyp(x,y,yp)=(Fe(x,y,yp+1e-6)-Fe(x,y,yp-1e-6))/(2e-6)
    d2Fdyp2(x,y,yp)=(Fe(x,y,yp+1e-6)-2Fe(x,y,yp)+Fe(x,y,yp-1e-6))/1e-12
    lambda, solve_el_bvp(Fe,dFdy,dFdyp,d2Fdyp2,a,b,ya,yb;n_grid=n_grid,tol=tol)
end

# ============================================================
# 可变端点问题
# ============================================================
"可变端点: y(a) 自由 → ∂F/∂y''|_{x=a}=0, 类似 y(b)"
function variable_left_endpoint_condition(F::Function, dFdyp::Function, a::Float64, y_a::Float64, yp_a::Float64)
    dFdyp(a, y_a, yp_a)
end

"横截条件 (端点沿曲线移动): (F-y''∂F/∂y'')δx + (∂F/∂y'')δy = 0"
function transversality_diff(F::Function, dFdyp::Function, x_end::Float64, y_end::Float64, yp_end::Float64, dx::Float64, dy::Float64)
    (F(x_end,y_end,yp_end)-yp_end*dFdyp(x_end,y_end,yp_end))*dx + dFdyp(x_end,y_end,yp_end)*dy
end

# ============================================================
# 多点边值 (Multiple-point BVP)
# ============================================================
"具有内部点的边值问题"
function multi_point_el_system(F::Function, a::Float64, b::Float64, interior_pts::Vector{Float64}, ya::Float64, yb::Float64; n_grid=500)
    # 将区间分段求解，内部点满足连续性+角条件
    segments=[[a]; interior_pts; [b]]
    m=length(segments)-1; solutions=[]
    for i in 1:m; push!(solutions, nothing); end  # placeholder
    solutions
end

# ============================================================
# 不等式约束
# ============================================================
"障碍问题: min J[y] s.t. y(x) ≥ ψ(x) (Signorini问题)"
function obstacle_problem_residual(F::Function, psi::Function, y::Function, yp::Function, x::Float64)
    if y(x) > psi(x)
        return 0.0  # 自由区域
    else
        return max(0.0, -(functional_derivative_at_x(F, y, yp, x)))  # 接触区域
    end
end

function functional_derivative_at_x(F::Function, y::Function, yp::Function, x::Float64; dx=1e-4)
    dFdy_val=(F(x+dx,y(x+dx),yp(x+dx))-F(x-dx,y(x-dx),yp(x-dx)))/(2dx)
    dFdyp_val=(F(x+dx,y(x+dx),yp(x+dx))-F(x-dx,y(x-dx),yp(x-dx)))/(2dx)
    d_dx_term=(dFdyp_val-dFdyp_val)/dx  # placeholder
    dFdy_val-d_dx_term
end

# ============================================================
# Bolza 问题 (Lagrange项 + Mayer项)
# ============================================================
"Bolza 问题: min φ(y(a),y(b)) + ∫_a^b F(x,y,y'')dx"
function bolza_boundary_condition(phi::Function, dFdyp::Function, a::Float64, b::Float64, y::Function, yp::Function)
    # 自然边界: ∂φ/∂y(a) + ∂F/∂y''|_a = 0, ∂φ/∂y(b) - ∂F/∂y''|_b = 0
    dphi_dya_val=0.0  # 需要 phi 的梯度信息
    dFdyp_a=dFdyp(a,y(a),yp(a))
    dphi_dya_val + dFdyp_a
end

export isoperimetric_lagrange_multiplier, dido_solution
export lagrange_multiplier_functional, solve_constrained_el_bvp
export variable_left_endpoint_condition, transversality_diff
export multi_point_el_system, obstacle_problem_residual, bolza_boundary_condition

File: mini-variational-principle/src/constrained.c
/**
 * constrained.c -- Constrained Variational Problems (L4-L6)
 */
#include "constrained.h"
#include "euler_lagrange.h"
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <float.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

/* Bisection root-finding helper */
static double bisect_root(double (*f)(double), double lo, double hi,
                          double tol, int max_iter) {
    double flo = f(lo), fhi = f(hi);
    for (int i = 0; i < max_iter; i++) {
        if (fabs(flo) < tol) return lo;
        if (fabs(fhi) < tol) return hi;
        double mid = 0.5*(lo+hi), fmid = f(mid);
        if (fabs(fmid) < tol) return mid;
        if (flo*fmid < 0.0) { hi=mid; fhi=fmid; }
        else { lo=mid; flo=fmid; }
        if (fabs(hi-lo) < tol) break;
    }
    return 0.5*(lo+hi);
}

double isoperimetric_lagrange_multiplier(
    LagrangianDensity F, LagrangianDensity G,
    double c, double a, double b,
    double ya, double yb, double lambda_guess)
{
    (void)lambda_guess;
    /* Constraint: integral G(x,y,y') dx = c */
    /* Search for lambda such that constraint is satisfied */
    double constraint_func(double lambda) {
        double dx = (b-a)/500.0;
        double y = ya, yp = (yb-ya)/(b-a);
        double K = 0.0;
        for (int i = 0; i < 500; i++) {
            double x = a + (double)i*dx;
            K += G(x, y, yp) * dx;
            /* Simple Euler step for modified Lagrangian F + lambda*G */
            double Fy = (F(x+1e-6,y,yp) + lambda*G(x+1e-6,y,yp) -
                         F(x-1e-6,y,yp) - lambda*G(x-1e-6,y,yp)) / (2e-6);
            yp += dx * Fy;
            y  += dx * yp;
        }
        return K - c;
    }
    return bisect_root(constraint_func, -100.0, 100.0, 1e-8, 60);
}

void dido_solution(double L, int n, double *x_vals, double *y_vals, double *R_out)
{
    double R = L / (2.0 * M_PI);
    for (int i = 0; i < n; i++) {
        double theta = 2.0 * M_PI * (double)i / (double)n;
        if (x_vals) x_vals[i] = R * cos(theta);
        if (y_vals) y_vals[i] = R * sin(theta);
    }
    if (R_out) *R_out = R;
}

void lagrange_multiplier_functional(
    LagrangianDensity F,
    LagrangianDensity *constraints_G, double *constraints_c,
    int m,
    double a, double b, double ya, double yb,
    double *lambdas_out)
{
    if (!lambdas_out || m < 1) return;
    /* Placeholder: initialize lambdas to zero */
    for (int i = 0; i < m; i++) lambdas_out[i] = 0.0;
    (void)F; (void)constraints_G; (void)constraints_c;
    (void)a; (void)b; (void)ya; (void)yb;
}

double solve_constrained_el_bvp(
    LagrangianDensity F, LagrangianDensity G,
    double c, double a, double b,
    double ya, double yb, int n_grid, double tol)
{
    double lambda = isoperimetric_lagrange_multiplier(F, G, c, a, b, ya, yb, 0.0);
    /* Build augmented Lagrangian */
    /* This function signature is simplified; full implementation would
     * require derivative functions for F+lambda*G */
    (void)n_grid; (void)tol;
    return lambda;
}

double variable_left_endpoint_condition(
    LagrangianDensity F, LagrangianDensity dF_dyp,
    double a, double y_a, double yp_a)
{
    (void)F; (void)a; (void)y_a;
    return dF_dyp ? dF_dyp(a, y_a, yp_a) : 0.0;
}

double transversality_diff(
    LagrangianDensity F, LagrangianDensity dF_dyp,
    double x_end, double y_end, double yp_end,
    double dx, double dy)
{
    if (!F || !dF_dyp) return 0.0;
    return (F(x_end,y_end,yp_end) - yp_end*dF_dyp(x_end,y_end,yp_end))*dx
           + dF_dyp(x_end,y_end,yp_end)*dy;
}

double multi_point_el_system(
    LagrangianDensity F,
    double a, double b,
    const double *interior_pts, int n_interior,
    double ya, double yb, int n_grid)
{
    (void)F; (void)a; (void)b; (void)interior_pts;
    (void)n_interior; (void)ya; (void)yb; (void)n_grid;
    /* Multi-point BVP: segment-wise solution with corner conditions */
    return 0.0;
}

double obstacle_problem_residual(
    LagrangianDensity F, double (*psi)(double),
    const double *y, const double *yp,
    double x, int idx, double dx)
{
    if (!F || !psi || !y || !yp) return 0.0;
    double psix = psi(x);
    if (y[idx] > psix) {
        return 0.0; /* Free region */
    } else {
        /* Contact region: compute EL residual */
        double Fy = (F(x+1e-6, y[idx], yp[idx]) -
                     F(x-1e-6, y[idx], yp[idx])) / (2e-6);
        double Fyp_p = (F(x+dx, y[idx+1], yp[idx+1]) -
                        F(x-dx, y[idx-1], yp[idx-1])) / (2.0*dx);
        return (Fy - Fyp_p > 0) ? Fy - Fyp_p : 0.0;
    }
}

double bolza_boundary_condition(
    double (*phi)(double, double),
    LagrangianDensity dF_dyp,
    double a, double b,
    const double *y, const double *yp, int n_nodes)
{
    if (!dF_dyp || !y || !yp || n_nodes < 2) return 0.0;
    (void)phi; (void)b;
    /* dphi/dy(a) + dF/dy'|_a = 0 */
    return dF_dyp(a, y[0], yp[0]);
}

double holonomic_constraint_el(
    double (*constraint)(double, double),
    double (*dconstraint_dy)(double, double),
    double x, double y, double yp, double lambda)
{
    if (!constraint || !dconstraint_dy) return 0.0;
    (void)x; (void)yp;
    /* Lagrange multiplier term: lambda * dg/dy */
    return lambda * dconstraint_dy(x, y);
}

double penalty_method_el(
    LagrangianDensity F,
    double (*constraint)(double, double),
    double x, double y, double yp,
    double penalty_param)
{
    if (!F || !constraint) return 0.0;
    double g = constraint(x, y);
    return F(x, y, yp) + 0.5 * penalty_param * g * g;
}

double augmented_lagrangian_el(
    LagrangianDensity F,
    double (*constraint)(double, double),
    double (*dconstraint_dy)(double, double),
    double x, double y, double yp,
    double lambda, double rho)
{
    if (!F || !constraint || !dconstraint_dy) return 0.0;
    double g = constraint(x, y);
    return F(x, y, yp) + lambda * g + 0.5 * rho * g * g;
}

double constraint_violation_norm(
    double (*constraint)(double, double),
    const double *x, const double *y, int n_nodes)
{
    if (!constraint || !x || !y || n_nodes < 2) return 0.0;
    double sum = 0.0;
    for (int i = 0; i < n_nodes; i++) {
        double g = constraint(x[i], y[i]);
        sum += g * g;
    }
    return sqrt(sum / (double)n_nodes);
}

bool verify_constraint_satisfaction(
    double (*constraint)(double, double),
    const double *x, const double *y, int n_nodes, double tol)
{
    return constraint_violation_norm(constraint, x, y, n_nodes) < tol;
}

int active_set_el(
    double (*constraint)(double, double),
    const double *x, const double *y, int n_nodes,
    double tol, int *active_set)
{
    if (!constraint || !x || !y || !active_set || n_nodes < 1) return 0;
    int count = 0;
    for (int i = 0; i < n_nodes; i++) {
        if (fabs(constraint(x[i], y[i])) < tol) {
            active_set[count++] = i;
        }
    }
    return count;
}
